ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java

```diff
@@ -705,52 +751,64 @@ int getSize() {
-      for (byte[] recoveryInfo : messageStoreRecoveryInfoList) {
+      for (HardDeletePersistItem item : items) {
         /* First write the size of the recoveryInfo */
-        dataOutputStream.writeInt(recoveryInfo.length);
+        dataOutputStream.writeInt(item.messagesStoreRecoveryInfo.length);
 
         /* Now, write the recoveryInfo */
-        dataOutputStream.write(recoveryInfo);
+        dataOutputStream.write(item.messagesStoreRecoveryInfo);
       }
 
       return outStream.toByteArray();
     }
 
     /**
-     * Prunes entries in the range from the start up to, but excluding, the entry with the passed in key.
+     * Prunes entries in the range from the start up to, but excluding, the entry with the passed in token.
+     * Token passed to this method has to be a indexed based one.
      */
-    void pruneTill(StoreKey storeKey) {
-      Iterator<BlobReadOptions> blobReadOptionsListIterator = blobReadOptionsList.iterator();
-      Iterator<byte[]> messageStoreRecoveryListIterator = messageStoreRecoveryInfoList.iterator();
-      while (blobReadOptionsListIterator.hasNext()) {
-      /* Note: In the off chance that there are multiple presence of the same key in this range due to prior software
-         bugs, note that this method prunes only till the first occurrence of the key. If it so happens that a
-         later occurrence is the one really associated with this token, it does not affect the safety.
-         Persisting more than what is required is okay as hard deleting a blob is an idempotent operation. */
-        messageStoreRecoveryListIterator.next();
-        if (blobReadOptionsListIterator.next().getMessageInfo().getStoreKey().equals(storeKey)) {
-          break;
+    void pruneTill(StoreFindToken token) {
+      Iterator<HardDeletePersistItem> itemsIterator = items.iterator();
+      while (itemsIterator.hasNext()) {
+        HardDeletePersistItem item = itemsIterator.next();
+        if (item.startTokenForBlobReadOptions.getType() == FindTokenType.Uninitialized) {
+          itemsIterator.remove();
         } else {
-          blobReadOptionsListIterator.remove();
-          messageStoreRecoveryListIterator.remove();
+          if (compareTwoTokens(item.startTokenForBlobReadOptions, token) >= 0) {
+            break;
+          } else {
+            itemsIterator.remove();
+          }
         }
       }
     }
 
-    private List<BlobReadOptions> getBlobReadOptionsList() {
-      return blobReadOptionsList;
+    /**
+     * Compare two StoreFindTokens and return the result as an integer like compareTo interface.
+     * These two tokens have to be IndexBased tokens.
+     * @param token1 The first token to compare.
+     * @param token2 The second tokent to compare.
+     * @return 0 means they are equal. negative number means token1 is less than token2. postive number means the opposite.
+     */
+    int compareTwoTokens(StoreFindToken token1, StoreFindToken token2) {
```

Comment:
The method looks good, minor suggestion is to align with your comment and do a token type check at the very beginning. (Reject Journal based token)

Reply:
good catch, updated.